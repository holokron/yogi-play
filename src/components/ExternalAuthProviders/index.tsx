import * as React from 'react'
import { UncontrolledAlert, Button, ButtonGroup, Col, Row } from 'reactstrap'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

export interface Props {
    googleLoginLabel: string
    facebookLoginLabel: string
    onLoginWithGoogle(): Promise<any>
    onLoginWithFacebook(): Promise<any>
}

export interface State {
    error: boolean
    isGoogleLoading: boolean
    isFacebookLoading: boolean
}

export default class ExternalAuthProviders extends React.PureComponent<Props, State> {
    constructor(props: Props) {
        super(props)

        this.state = {
            error: false,
            isGoogleLoading: false,
            isFacebookLoading: false,
        }

        this.handleGoogleLogin = this.handleGoogleLogin.bind(this)
        this.handleFacebookLogin = this.handleFacebookLogin.bind(this)
    }

    async handleGoogleLogin(event: React.MouseEvent<HTMLButtonElement>): Promise<void> {
        event.preventDefault()
        
        if (this.state.isGoogleLoading) {
            return
        }

        this.setState({
            error: false,
            isGoogleLoading: true,
        })

        try {
            await this.props.onLoginWithGoogle()
        } catch (error) {
            this.setState({
                error: true,
                isGoogleLoading: false,
            })
        }
    }

    async handleFacebookLogin(event: React.MouseEvent<HTMLButtonElement>): Promise<void> {
        event.preventDefault()
        
        if (this.state.isFacebookLoading) {
            return
        }

        this.setState({
            error: false,
            isFacebookLoading: true,
        })

        try {
            await this.props.onLoginWithFacebook()
        } catch (error) {
            this.setState({
                error: true,
                isFacebookLoading: false,
            })
        }
    }

    render () {
        const {
            state: {
                error,
                isGoogleLoading,
                isFacebookLoading,
            },
            handleGoogleLogin,
            handleFacebookLogin,
            props: {
                googleLoginLabel,
                facebookLoginLabel,
            },
        } = this

        return (
            <Row>
                {error && 
                    <Col md="12" className="mt-3">
                        <UncontrolledAlert color="danger">
                            Bład połaczenia lub konto o tym samym adresie email już istnieje.
                            Spróbuj zalogować się innym kontem.
                        </UncontrolledAlert>
                    </Col>
                }
                <Col md="12" className="mt-3">
                    <ButtonGroup className="btn-block">
                        <Button active color="primary">
                            <FontAwesomeIcon icon="google" size="lg" fixedWidth />
                        </Button>
                        <Button onClick={handleGoogleLogin} color="primary" className="btn-block">
                            {isGoogleLoading && <FontAwesomeIcon icon="spinner" pulse fixedWidth />}
                            &nbsp; {googleLoginLabel}
                        </Button>
                    </ButtonGroup>
                </Col>
                <Col md="12" className="mt-3">
                    <ButtonGroup className="btn-block">
                        <Button active color="primary">
                            <FontAwesomeIcon icon="facebook" size="lg" fixedWidth />
                        </Button>
                        <Button onClick={handleFacebookLogin} color="primary" className="btn-block">
                            {isFacebookLoading && <FontAwesomeIcon icon="spinner" pulse fixedWidth />}
                            &nbsp; {facebookLoginLabel}
                        </Button>
                    </ButtonGroup>
                </Col>
            </Row>
        )
    }
}