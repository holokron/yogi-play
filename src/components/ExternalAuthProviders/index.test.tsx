import * as React from 'react'
import * as renderer from 'react-test-renderer'
import ExternalAuthProviders from '.'

describe('@components/ExternalAuthProviders', () => {
    it('renders correctly', () => {
        const tree = renderer.create(
            <ExternalAuthProviders
                googleLoginLabel="Login with Google"
                facebookLoginLabel="Login with Facebook"
                onLoginWithGoogle={async () => {}}
                onLoginWithFacebook={async () => {}}
            />
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })
})