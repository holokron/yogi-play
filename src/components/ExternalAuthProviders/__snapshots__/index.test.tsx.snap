// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/ExternalAuthProviders renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="mt-3 col-md-12"
  >
    <div
      className="btn-block btn-group"
      role="group"
    >
      <button
        className="btn btn-primary active"
        onClick={[Function]}
        type={undefined}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-google fa-w-16 fa-fw fa-lg "
          data-icon="google"
          data-prefix="fab"
          role="img"
          style={Object {}}
          viewBox="0 0 488 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="btn-block btn btn-primary"
        onClick={[Function]}
        type="button"
      >
          
        Login with Google
      </button>
    </div>
  </div>
  <div
    className="mt-3 col-md-12"
  >
    <div
      className="btn-block btn-group"
      role="group"
    >
      <button
        className="btn btn-primary active"
        onClick={[Function]}
        type={undefined}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-facebook fa-w-14 fa-fw fa-lg "
          data-icon="facebook"
          data-prefix="fab"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 56.7v398.5c0 13.7-11.1 24.7-24.7 24.7H309.1V306.5h58.2l8.7-67.6h-67v-43.2c0-19.6 5.4-32.9 33.5-32.9h35.8v-60.5c-6.2-.8-27.4-2.7-52.2-2.7-51.6 0-87 31.5-87 89.4v49.9h-58.4v67.6h58.4V480H24.7C11.1 480 0 468.9 0 455.3V56.7C0 43.1 11.1 32 24.7 32h398.5c13.7 0 24.8 11.1 24.8 24.7z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="btn-block btn btn-primary"
        onClick={[Function]}
        type="button"
      >
          
        Login with Facebook
      </button>
    </div>
  </div>
</div>
`;
